Add nrf_coresight driver

karstenkoenig opened the pull request:
The coresight subsystem for the Nordic Haltium family of devices is a combination of different coresight modules that can be used in combination for logging, data tracing, instruction tracing etc.
This adds a driver and bindings for the coresight subsystem that powers and then applies a configuration based on a selected operating mode. It replaces the previous system where the configuration was applied by a closed-source firmware bit.

This also adds the needed pinctrl handling for trace pins.
Furthermore the existing samples and snippets are adjusted accordingly.

karstenkoenig committed:
drivers: pinctrl_nrf: Add coresight tpiu pins

Pinctrl needs to set the needed drive and direction of the pins. Also
this later allows automatically setting the clock bit for the traceclk
pin.

Signed-off-by: Karsten Koenig <karsten.koenig@nordicsemi.no>

karstenkoenig committed:
drivers: debub: coresight: Added coresight_nrf

Added driver and bindings for the coresight nrf submodule.
add integrated it for the nrf54h20.
The coresight subsystem is a combination of ARM Coresight peripherals
that get configured together to achieve a simplified configuration based
on a desired operating mode.

This also replaces the previous handling in the nrf54h20 soc.c which was
powering the subsystem up but not configuring it.

Signed-off-by: Karsten Koenig <karsten.koenig@nordicsemi.no>

karstenkoenig committed:
snippets: nordic-log-stm: Updated for coresight

Replaced the old tddconf with the full coresight driver that configures
the coresight peripherals locally on the running core.
Also fixed minor bug in the corresponding sample where messages were not
shown for the radio core.

Signed-off-by: Karsten Koenig <karsten.koenig@nordicsemi.no>

karstenkoenig committed:
drivers: misc: nordic_vpr_launcher: Init order

Make sure the nordic_vpr_launcher gets started after the coresight
driver if that is present.

Signed-off-by: Karsten Koenig <karsten.koenig@nordicsemi.no>

karstenkoenig committed:
snippets: nordic-log-stm-tpiu-dict: Added

Added a new convience snippet to redirect logs to STM and then sink them
to TPIU where they can be captured by a trace probe.

Signed-off-by: Karsten Koenig <karsten.koenig@nordicsemi.no>

nordic-krch reviewed (commented):
Looks ok. One more thing that might be added is early logging on app. On sdfw there was a RAM buffer (`CONFIG_LOG_FRONTEND_STMESP_EARLY_BUF_SIZE`!=0) that was collecting logging data and flushing it to STM when system notified that ETR is ready. using `log_frontend_stmesp_etr_ready()`. Now it should be done on app but it can be added later as well.

karstenkoenig commented on drivers/misc/coresight/nrf_coresight.c:
```diff
@@ -0,0 +1,392 @@
+		}
+	}
+
+	err = ironside_se_tdd_configure(IRONSIDE_SE_TDD_CONFIG_ON_DEFAULT);
```
Yes good idea, you are right there is quite a bit in the coresight subsystem that'd be nice to use without having to handle a full blown mode (beyond power up tdd and pinctrl) in the coresight driver. This will also be convenient to use with ETM, as there coresight can be configured externally from a debugger.

karstenkoenig replied:
I added `CORESIGHT_NRF_MODE_UNCONFIGURED` to indicate coresight subsystem is active but unconfigured, so that pinctrl and tdd power are handled, but the coresight peripherals themselves remain at reset values

masz-nordic reviewed (changes requested):
Driver should be using HAL and fixups should be added to the nrfx.

karstenkoenig commented:
> Driver should be using HAL and fixups should be added to the nrfx.

You mean the nrf hal? There is no nrf hal for the coresight peripherals, these aren't nordic IPs, they come from ARM.

How about I add a coresight.h header where the coresight peripherals are defined in an MDK like fashion but no reference to NRF types but instead ARM_<CoresightModule>_Type? Could I then write straight to such `ARM_` structs in the coresight_nrf.c driver or would I still need to add an extra hal layer to set fields?

masz-nordic commented:
Sounds good, then the driver could be vendor-agnostic - `coresight_arm.c`.

karstenkoenig commented:
> for STMESP I did https://github.com/zephyrproject-rtos/zephyr/blob/8d508a2f94970561b696a87feb35132c4a9dcc1a/include/zephyr/drivers/misc/coresight/stmesp.h.

Yes I am looking into something like this for the coresight peripherals. I was going to do something without structs for the register map but OFFSET definitions but this actually does look nicer, so I'll change it to that.

> Are the funnels architecture in TDD Nordic specific?

The funnels aren't nordic specific, but the way the whole trace and debug domain is setup is highly nordic specific, so the way all the different funnel and replicators link together various trace data sources and sinks. So I still plan to have a coresight_nrf that just configures all of those in one go and assume this design will hold for other users as well.

karstenkoenig commented:
@masz-nordic @nordic-krch I have gone through a few iterations with coresight_arm, but figured a simple hal like thing with offsets and msk/pos definitions plus lock/unlock functions is as good as it gets in terms of common functionality. I don't know how other platforms integrated the coresight peripherals and don't want to impose a big untested abstraction layer for coresight_arm so limited it what the coresight_nrf driver is using.

karstenkoenig commented on drivers/misc/nordic_vpr_launcher/Kconfig:
```diff
@@ -18,8 +18,10 @@ source "subsys/logging/Kconfig.template.log_config"
 	default 0
 	help
 	  The init priority of the VPR coprocessor launcher.
+	  Needs to be higher than the init priority of NRF coresight if enabled.
```
removed the comment

tmon-nordic reviewed (dismissed):
I have verified that `-S nordic-log-stm` works as expected when using sysbuild to build project that uses both App core and FLPR.

karstenkoenig committed:
boards: nordic: nrf54h20dk: ETM in JLinkScript

The JLinkScript originally configured more than needed which reduces
readability, so stripped it down to the minimum.
At the same time improved behavior under reset and added ETM to the
radiocore JLinkScript.

Signed-off-by: Karsten Koenig <karsten.koenig@nordicsemi.no>

karstenkoenig commented on drivers/misc/coresight/coresight_nrf.c:
```diff
@@ -1,8 +1,9 @@
 /*
- * Copyright (c) 2025 Nordic Semiconductor ASA.
+ * Copyright (c) 2025 Nordic Semiconductor ASA
  *
- * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
+ * SPDX-License-Identifier: Apache-2.0
  */
+
```
done

karstenkoenig commented:
Rebased to avoid https://github.com/zephyrproject-rtos/zephyr/issues/96610 which breaks some twister job here but is unrelated

karstenkoenig commented on drivers/misc/coresight/Kconfig:
```diff
@@ -111,3 +111,72 @@ config NRF_ETR_SHELL_ASYNC_RX_BUFFER_COUNT
+	  CoreSight System Trace Macrocell (STM) trace support.
+
+if CORESIGHT_NRF
+config CORESIGHT_NRF_ATB_TRACE_ID_STM_GLOBAL
```
added

karstenkoenig commented on drivers/misc/coresight/Kconfig:
```diff
@@ -111,3 +111,72 @@ config NRF_ETR_SHELL_ASYNC_RX_BUFFER_COUNT
+config CORESIGHT_NRF_TDD_CLOCK_FREQUENCY
+	int
+	default 320000000 if SOC_SERIES_NRF54HX
+	default 400000000
```
It can, added it. It's not really a clock the user has control over so I added it as a fixed clock and not as a variant of the nrf hsflls

karstenkoenig commented on drivers/misc/coresight/Kconfig:
```diff
@@ -111,3 +111,72 @@ config NRF_ETR_SHELL_ASYNC_RX_BUFFER_COUNT
+	default 512
+
+config CORESIGHT_NRF_STM_HWEVENTS
+	bool "CoreSight STM: Enable hardware events"
```
no particular reason, good suggestion so did as you suggested

karstenkoenig commented:
@nordicjm  Thanks for your feedback, implemented all of your points

nordicjm reviewed (commented):
minor nit. then is OK

karstenkoenig commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-pinctrl.dtsi:
```diff
@@ -138,4 +138,16 @@
+			nordic,drive-mode = <NRF_DRIVE_H0H1>;
+		};
+	};
+
```
done

asmellby reviewed (changes requested):
Shouldn't this be added to `drivers/debug/` instead of `drivers/misc/`?

That seemed to be the Arch WG consensus in https://github.com/zephyrproject-rtos/zephyr/pull/88035 .

cc @carlescufi

karstenkoenig commented:
> Shouldn't this be added to `drivers/debug/` instead `drivers/misc/`?

Oh I didn't know about this debug drivers place, it absolutely looks like a better spot! Thanks for catching this.
Especially as you in your PR also discussed the idea of general 'trace init' that lots of platforms have in various forms.

Should I just move the whole coresight folder as suggested in my PR into `debug` including the existing `nrf_etr.c`? Should coresight be retained as a subfolder or should it just be straight in `drivers/debug`?

asmellby commented:
I think putting it directly in `drivers/debug` is fine, we already didn't add a subdirectory for the one driver that's currently in there (the Silicon Labs PTI driver). And yeah, I think it would be great if you also move the `nrf_etr.c`, but I guess @nordic-krch should chime in on that.

nordic-krch commented:
> @nordic-krch Any objection to me moving nrf_etr.c from misc into debug?

@karstenkoenig no objectionss

karstenkoenig committed:
drivers: debug: Moved nrf_etr from misc

Moved the nrf_etr driver from the drive/misc folder into the recently
established driver/debug folder where it is a better fit. Moved the
associated files such as bindings and headers accordingly as well.

Signed-off-by: Karsten Koenig <karsten.koenig@nordicsemi.no>

karstenkoenig commented:
For all the other reviewers I didn't change the coresight_nrf part except for moving it to drivers/debug. I had to rebase on top of main as there would have been a conflict in the pinctrl handling

asmellby reviewed (dismissed):
Looks good to me.
You may want to add a migration guide entry for the renamed ETR Kconfig options?

tmon-nordic reviewed (dismissed):
I didn't review the code, but tried that `-S nordic-log-stm` works with nRF54H20 FLPR. In separate project I have confirmed that I can see instructions in Ozone 3.40c Instruction Trace window  on nRF54H20 App core.

karstenkoenig committed:
doc: migration-guide: Moved nrf_etr

Migration guide for the move of drivers/misc/nrf_etr to
drivers/debug/debug_nrf_etr.

Signed-off-by: Karsten Koenig <karsten.koenig@nordicsemi.no>

kartben commented:
@nordic-krch there were already 3 assignees on this, can you please document and elaborate on why you self-assigned?

nordic-krch commented:
@kartben I was the initial author of coresight/STM Nordic support in Zephyr so most of modified files are were created by me. I don't think that @anangl @masz-nordic @bjarki-andreasen have much expertise in that area so I would not expect a review from them.

nordic-krch commented:
I should probably have added myself as the maintainer before.
